cpp/input/event.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include "event.h"
+
+#include <functional>
+
+namespace openage {
+namespace input {
+
+class_code_t::class_code_t(event_class cl, code_t code)
+	:
+	eclass(cl),
+	code(code) {
+}
+
+
+std::vector<event_class> class_code_t::get_classes() const {
+	std::vector<event_class> result;
+
+	// use event_base to traverse up the class tree
+	event_class ec = this->eclass;
+	result.push_back(ec);
+	while (event_base.count(ec) > 0) {
+		ec = event_base.at(ec);
+		result.push_back(ec);
+	}
+	return result;
+}
+
+
+bool class_code_t::has_class(const event_class &ec) const {
+	for (auto c : this->get_classes()) {
+		if (c == ec) {
+			return true;
+		}
+	}
+	return false;
+}
+
+
+bool operator ==(class_code_t a, class_code_t b) {
+	return a.eclass == b.eclass && a.code == b.code;
+}
+
+
+Event::Event(event_class cl, code_t code, modset_t mod)
+	:
+	cc(cl, code),
+	mod(mod) {}
+
+
+char Event::as_char() const {
+	return (cc.code & 0xff);
+}
+
+
+std::string Event::info() const {
+	std::string result;
+	result += "[Event: ";
+	result += std::to_string(static_cast<int>(this->cc.eclass));
+	result += ", ";
+	result += std::to_string(this->cc.code);
+	result += ", ";
+	result += this->as_char();
+	result += " (";
+	result += std::to_string(this->mod.size());
+	result += ")]";
+	return result;
+}
+
+
+bool Event::operator ==(const Event &other) const {
+	return this->cc == other.cc && this->mod == other.mod;
+}
+
+
+int event_hash::operator()(const Event &e) const {
+	return class_code_hash()(e.cc); // ^ std::hash<modset_t>()(e.mod) * 3664657;
```

Comment:
Why do we need `Event` to be hashable at all? At first glance, this seems like something that should be unhashable by design.